index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,198 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"regexp"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+// simpleTermsDictionary uses two-level map to model a terms dictionary.
+// i.e. fieldName -> fieldValue -> postingsList
+type simpleTermsDictionary struct {
+	fieldNamesLock sync.RWMutex
+	fieldNames     map[string]*fieldValuesMap
+
+	opts Options
+}
+
+func newSimpleTermsDictionary(opts Options) termsDictionary {
+	return &simpleTermsDictionary{
+		fieldNames: make(map[string]*fieldValuesMap, opts.InitialCapacity()),
+		opts:       opts,
+	}
+}
+
+func (t *simpleTermsDictionary) Insert(field doc.Field, i segment.DocID) error {
+	fieldName := string(field.Name)
+	fieldValues := t.getOrAddFieldName(fieldName)
+	fieldValue := string(field.Value)
+	return fieldValues.addDocIDForValue(fieldValue, i)
+}
+
+func (t *simpleTermsDictionary) Fetch(
+	fieldName []byte,
+	fieldValueFilter []byte,
+	isRegexp bool,
+) (segment.PostingsList, error) {
+	// check if we know about the field name
+	t.fieldNamesLock.RLock()
+	fieldValues, ok := t.fieldNames[string(fieldName)]
+	t.fieldNamesLock.RUnlock()
+	if !ok {
+		// not an error to not have any matching values
+		return nil, nil
+	}
+
+	// get postingList(s) for the given value.
+	lists, err := fieldValues.fetchLists(fieldValueFilter, isRegexp)
+	if err != nil {
+		return nil, err
+	}
+
+	// union all the postingsList(s)
+	unionedList := t.opts.PostingsListPool().Get()
+	for _, ids := range lists {
+		unionedList.Union(ids)
+	}
+
+	return unionedList, nil
+}
+
+func (t *simpleTermsDictionary) getOrAddFieldName(fieldName string) *fieldValuesMap {
+	// cheap read lock to see if it already exists
+	t.fieldNamesLock.RLock()
+	fieldValues, ok := t.fieldNames[fieldName]
+	t.fieldNamesLock.RUnlock()
+	if ok {
+		return fieldValues
+	}
+
+	// acquire write lock and create
+	t.fieldNamesLock.Lock()
```

Comment:
Probably cheap enough to `defer ...Unlock()` here.

Reply:
Yeah just think it's a small enough function that it doesn't hurt readability to not use a defer here. Only one exit point in the if right after this needs the extra unlock call.